File: apps/mobile/src/domain/content-ids.ts
import { EXERCISE_ID_SUFFIX, type ExerciseType } from "./exercise-types";

// Versões anteriores do app usavam `letter-a` e `A-listen`; o formato canônico é o do servidor.
const LEGACY_LESSON = /^letter-([a-z])$/;
const EXERCISE = /^(?:exercise-)?([A-Z])-(listen|recognize|find|find_in_word|complete-word)$/;

export const lessonIdFor = (letter: string) => `lesson-${letter.toUpperCase()}`;
export const exerciseIdFor = (letter: string, type: ExerciseType) => `exercise-${letter.toUpperCase()}-${EXERCISE_ID_SUFFIX[type]}`;

export function canonicalLessonId(id: string): string {
  const match = LEGACY_LESSON.exec(id);
  return match ? lessonIdFor(match[1]) : id;
}

export function canonicalExerciseId(id: string): string {
  const match = EXERCISE.exec(id);
  return match ? `exercise-${match[1]}-${match[2] === "find_in_word" ? "find" : match[2]}` : id;
}

File: apps/api/app/content_ids.py
"""IDs canônicos de conteúdo.

Versões anteriores do app usavam `letter-a` e `A-listen`, e seeds novos geravam `exercise-A-find_in_word`.
Unidades de letra: `lesson-A`; itens de letra: `exercise-A-<sufixo do contrato>`; itens de trilha temática: `item-<unidade>-<n>`.
"""
import re

from .content_types import EXERCISE_ID_SUFFIX, ExerciseType

_LEGACY_LESSON = re.compile(r"^letter-([a-z])$")
_EXERCISE = re.compile(r"^(?:exercise-)?([A-Z])-(listen|recognize|find|find_in_word|complete-word|syllable)$")


def lesson_id_for(letter: str) -> str:
    return f"lesson-{letter.upper()}"


def exercise_id_for(letter: str, kind: str) -> str:
    return f"exercise-{letter.upper()}-{EXERCISE_ID_SUFFIX[ExerciseType(kind)]}"


def item_id_for(unit_id: str, index: int) -> str:
    return f"item-{unit_id}-{index}"


def canonical_lesson_id(value: str) -> str:
    match = _LEGACY_LESSON.match(value)
    return lesson_id_for(match.group(1)) if match else value


def canonical_exercise_id(value: str) -> str:
    match = _EXERCISE.match(value)
    if not match:
        return value
    return f"exercise-{match.group(1)}-{'find' if match.group(2) == 'find_in_word' else match.group(2)}"
